perception/autoware_tensorrt_bevdet/src/preprocess_plugin.cu

```diff
@@ -0,0 +1,347 @@
+  cudaStream_t stream) noexcept
+{
+  int n_img = inputDesc[0].dims.d[0];
+  int src_img_h = inputDesc[0].dims.d[2];
+  int src_img_w = inputDesc[0].dims.d[3] * 4;
+
+  int dst_img_h = outputDesc[0].dims.d[2];
+  int dst_img_w = outputDesc[0].dims.d[3];
+
+  int src_row_step = src_img_w;
+  int dst_row_step = dst_img_w;
+
+  int src_img_step = src_img_w * src_img_h * 3;
+  int dst_img_step = dst_img_w * dst_img_h * 3;
+
+  float offset_h = m_.crop_h / m_.resize_radio;
+  float offset_w = m_.crop_w / m_.resize_radio;
+
+  dim3 grid(DIVUP(dst_img_h * dst_img_w * n_img, NUM_THREADS));
+  dim3 block(NUM_THREADS);
+
+  switch (int(outputDesc[0].type)) {
+    case int(DataType::kFLOAT):
+      // float
+      preprocess_kernel<<<grid, block, 0, stream>>>(
+        reinterpret_cast<const uint8_t *>(inputs[0]), reinterpret_cast<float *>(outputs[0]),
+        src_row_step, dst_row_step, src_img_step, dst_img_step, src_img_h, src_img_w,
+        m_.resize_radio, m_.resize_radio, offset_h, offset_w,
+        reinterpret_cast<const float *>(inputs[1]), reinterpret_cast<const float *>(inputs[2]),
+        dst_img_h, dst_img_w, n_img);
+      break;
+    case int(DataType::kHALF):
+      // half
+      preprocess_kernel<<<grid, block, 0, stream>>>(
+        reinterpret_cast<const uint8_t *>(inputs[0]), reinterpret_cast<__half *>(outputs[0]),
+        src_row_step, dst_row_step, src_img_step, dst_img_step, src_img_h, src_img_w,
+        m_.resize_radio, m_.resize_radio, offset_h, offset_w,
+        reinterpret_cast<const float *>(inputs[1]), reinterpret_cast<const float *>(inputs[2]),
+        dst_img_h, dst_img_w, n_img);
+
+      break;
+    default:  // should NOT be here
+      RCLCPP_ERROR(rclcpp::get_logger("PreprocessPlugin"), "\tUnsupport datatype!");
+  }
+  return 0;
+}
+
+void PreprocessPlugin::destroy() noexcept
+{
+  delete this;
+  return;
+}
+
+int32_t PreprocessPlugin::initialize() noexcept
+{
+  return 0;
+}
+
+void PreprocessPlugin::terminate() noexcept
+{
+  return;
+}
+
+size_t PreprocessPlugin::getSerializationSize() const noexcept
+{
+  return sizeof(m_);
+}
+
+void PreprocessPlugin::serialize(void * buffer) const noexcept
+{
+  memcpy(buffer, &m_, sizeof(m_));
+  return;
+}
+
+void PreprocessPlugin::setPluginNamespace(const char * pluginNamespace) noexcept
+{
+  namespace_ = pluginNamespace;
+  return;
+}
+
+const char * PreprocessPlugin::getPluginNamespace() const noexcept
+{
+  return namespace_.c_str();
+}
+
+const char * PreprocessPlugin::getPluginType() const noexcept
+{
+  return PRE_PLUGIN_NAME;
+}
+
+const char * PreprocessPlugin::getPluginVersion() const noexcept
+{
+  return PRE_PLUGIN_VERSION;
+}
+
+void PreprocessPlugin::attachToContext(
+  cudnnContext * contextCudnn, cublasContext * contextCublas, IGpuAllocator * gpuAllocator) noexcept
+{
+  return;
+}
+
+void PreprocessPlugin::detachFromContext() noexcept
+{
+  return;
+}
+
+// class PreprocessPluginCreator
```

Comment:
Similar to the plugin class, consider wrapping the `PreprocessPluginCreator` class within the same namespace used for the plugin class to maintain consistency and avoid naming clashes.